miner/block_builder.go

```diff
@@ -0,0 +1,192 @@
+		newTrans:make(chan *state.Transaction),
+		hareResult:make(chan HareResult),
+		transactionQueue:make([]SerializableTransaction,0,10),
+		hareRes:nil,
+		mu:sync.Mutex{},
+		network:net,
+		weakCoinToss:weakCoin,
+		orphans:orph,
+		started: false,
+	}
+}
+
+func Transaction2SerializableTransaction(tx *state.Transaction) SerializableTransaction{
+	return SerializableTransaction{
+		AccountNonce:tx.AccountNonce,
+		Origin:tx.Origin,
+		Recipient:tx.Recipient,
+		Amount:tx.Amount.Bytes(),
+		Payload:tx.Payload,
+		GasLimit:tx.GasLimit,
+		Price:tx.Price.Bytes(),
+	}
+}
+
+func (t *BlockBuilder) Start() error{
+	if t.started {
+		return fmt.Errorf("already started")
+	}
+
+	t.started = true
+	go t.acceptBlockData()
+	return nil
+}
+
+func (t *BlockBuilder) Stop() error{
+	if !t.started {
+		return fmt.Errorf("already stopped")
+	}
+	t.stopChan <- struct{}{}
+	return nil
+}
+
+type HareResult interface {
+	GetResult() []mesh.BlockID
+}
+
+type WeakCoinProvider interface {
+	GetResult() bool
+}
+
+type OrphanBlockProvider interface {
+	GetOrphans() []mesh.BlockID
+}
+
+func (t *BlockBuilder) CommitHareResult(res HareResult) error{
+	if !t.started{
+		return fmt.Errorf("BlockBuilderStopped")
+	}
+	t.hareResult <- res
+	return nil
+}
+
+func (t *BlockBuilder) AddTransaction(nonce uint64, origin, destination common.Address, amount *big.Int) error{
+	if !t.started{
+		return fmt.Errorf("BlockBuilderStopped")
+	}
+	tx := state.NewTransaction(nonce,origin,destination,amount)
+	t.newTrans <- tx
+	return nil
+}
+
+func (t *BlockBuilder) resetAfterCreateBlock(){
+	t.hareRes = nil
+}
+
+func (t *BlockBuilder) createBlock(id mesh.LayerID, txs []SerializableTransaction) Block {
+	b := Block{
+		Id : mesh.BlockID(rand.Int63()),
+		LayerIndex:id,
+		Data:nil,
+		Coin: t.weakCoinToss.GetResult(),
+		Timestamp:time.Now(),
+		Txs : make([]SerializableTransaction, len(txs)),
+		HareResult : t.hareRes.GetResult(),
+		Orphans:t.orphans.GetOrphans(),
+	}
+	copy(b.Txs, txs)
+	t.resetAfterCreateBlock()
+
+	return b
+}
+func BlockAsBytes(block Block) ([]byte, error) {
+	var w bytes.Buffer
+	if _, err := xdr.Marshal(&w, &block); err != nil {
+		return nil, fmt.Errorf("error marshalling block ids %v", err)
+	}
+	return w.Bytes(), nil
+}
+
+func (t *BlockBuilder) acceptBlockData(){
+	for{
+		select {
+		case <-t.stopChan:
+			t.started = false
+			return
+
+		case id := <-t.beginRoundEvent:
+			txList := t.transactionQueue[:common.Min(len(t.transactionQueue), MaxTransactionsPerBlock)]
+			t.transactionQueue = t.transactionQueue[common.Min(len(t.transactionQueue), MaxTransactionsPerBlock):]
+			blk := t.createBlock(id, txList)
+			go func(){
+				bytes , err := BlockAsBytes(blk)
+				if err != nil {
+					log.Error("cannot serialize block %v", err)
+					return
+				}
+				t.network.Broadcast(meshSync.BlockProtocol,bytes)
+			}()
+
+		case tx := <- t.newTrans:
+			t.transactionQueue = append(t.transactionQueue,Transaction2SerializableTransaction(tx))
```

Comment:
its easy to put that in a select loop but we've already discussed mutexes are more efficient when projecting state

Reply:
this is a more flexible code, channels can be consumed from arious producers if we decide to, and will eliminate the need for functions that need to be locked with this mutex. also, if we decide the channel receiving txs should be buffered it will be more efficient than simple locking a mutex when adding a single value